src/amo/components/CollectionAddAddon/index.js

```diff
@@ -65,27 +70,51 @@ export class CollectionAddAddonBase extends React.Component<InternalProps> {
   };
 
   componentWillReceiveProps(props: InternalProps) {
-    const { hasAddonBeenAdded: hasAddonBeenAddedNew } = props;
-    const { hasAddonBeenAdded } = this.props;
-    if (hasAddonBeenAdded !== hasAddonBeenAddedNew) {
+    const {
+      hasAddonBeenAdded: hasAddonBeenAddedNew,
+      hasAddonBeenRemoved: hasAddonBeenRemovedNew,
+    } = props;
+    const { errorHandler, hasAddonBeenAdded, hasAddonBeenRemoved } = this.props;
+
+    const addStatusChanged = hasAddonBeenAdded !== hasAddonBeenAddedNew;
```

Comment:
The complexity of having to compare the previous value with the new value struck me as odd when I saw the first patch but it only just dawned on me as to what the problem is.

You're mixing two state managers. You're mixing `setUIState()` with the `collections` reducer. You should be able to accomplish all of this with just the `collections` reducer.

Rendering the notification should be as simple as checking the mapped state prop:
```diff
diff --git a/src/amo/components/CollectionAddAddon/index.js b/src/amo/components/CollectionAddAddon/index.js
index 317831e7e..b71bb903d 100644
--- a/src/amo/components/CollectionAddAddon/index.js
+++ b/src/amo/components/CollectionAddAddon/index.js
@@ -138,7 +138,13 @@ export class CollectionAddAddonBase extends React.Component<InternalProps> {
   };
 
   render() {
-    const { errorHandler, i18n, uiState } = this.props;
+    const {
+      errorHandler,
+      hasAddonBeenAdded,
+      hasAddonBeenRemoved,
+      i18n,
+      uiState,
+    } = this.props;
     const { addonAction } = uiState;
     const addonAdded = addonAction === addonAddedAction;
 
@@ -148,13 +154,13 @@ export class CollectionAddAddonBase extends React.Component<InternalProps> {
           errorHandler.renderError()
         ) : (
           <TransitionGroup className="CollectionAddAddon-noticePlaceholder">
-            {addonAction && (
+            {(hasAddonBeenAdded || hasAddonBeenRemoved) && (
               <CSSTransition
                 classNames="CollectionAddAddon-noticePlaceholder-transition"
                 timeout={MESSAGE_FADEOUT_TIME}
               >
-                <Notice type={addonAdded ? 'success' : 'generic'}>
-                  {addonAdded
+                <Notice type={hasAddonBeenAdded ? 'success' : 'generic'}>
+                  {hasAddonBeenAdded
                     ? i18n.gettext('Added to collection')
                     : i18n.gettext('Removed from collection')}
                 </Notice>
```

However, this means you also need to hide the notification by dispatching an action in a timeout. Something like this:

```diff
diff --git a/src/amo/components/CollectionAddAddon/index.js b/src/amo/components/CollectionAddAddon/index.js
index 317831e7e..0b0e2e226 100644
--- a/src/amo/components/CollectionAddAddon/index.js
+++ b/src/amo/components/CollectionAddAddon/index.js
@@ -71,11 +71,19 @@ export class CollectionAddAddonBase extends React.Component<InternalProps> {
 
   componentWillReceiveProps(props: InternalProps) {
     const {
+      dispatch,
       hasAddonBeenAdded: hasAddonBeenAddedNew,
       hasAddonBeenRemoved: hasAddonBeenRemovedNew,
+      setTimeout,
     } = props;
     const { errorHandler, hasAddonBeenAdded, hasAddonBeenRemoved } = this.props;
 
+    if (hasAddonBeenAdded) {
+      setTimeout(() => {
+        dispatch(hideAddonAddedNotification());
+      }, MESSAGE_FADEOUT_TIME);
+    }
+
     const addStatusChanged = hasAddonBeenAdded !== hasAddonBeenAddedNew;
     const removeStatusChanged = hasAddonBeenRemoved !== hasAddonBeenRemovedNew;
 
```

Both of these seem like simple changes but it would take quite a bit of redesign so I don't suggest you try and do it in this patch. You could do it in a future patch if you like but it would of course take time and effort. The net benefit is probably just that `componentWillReceiveProps()` would become simpler.

Anyway, just food for thought!

Reply:
Thanks @kumar303, you're right, that's much better and much simpler. Because of that, I'm actually inclined to fix it now even though it's a bit of work. Otherwise the "worse" approach I'm using will just become more technical debt and we have a big enough backlog of that already. I don't think it will be a *huge* amount of work, so I'll try to get it done in time for the next tag.